### PR TITLE
Added mainAxisExtent to SliverGridDelegateWithFixedCrossAxisCount and SliverGridDelegateWithMaxCrossAxisExtent

### LINKED ISSUE #71282: Proposal to add new SliverGridDelegate with fixed main axis extent instead of aspectRatio
I think it would be a great feature to be able to create a cell with a fixed extent.

I find this considerably easy to implement and a great advantage. Right now for GridView and SliverGridView the cell size can be defined by:
 -   SliverGridDelegateWithFixedCrossAxisCount
 -  SliverGridDelegateWithFixedCrossAxisExtent

In both, the main axis extent can only be defined by a ratio relative to the crossAxisExtent. This becomes difficult to use when you expect to have a card/tile with the same extent even if the window is resized.

|  Extent(Proposal) |  AspectRatio |
|---|---|
|| | 

The code needed is simple and it could be added as new constructor for both current delegates or create 2 new SliverGridDelegate.

```dart
class SliverGridDelegateWithFixedCrossAxisCountAndMainAxisExtent extends SliverGridDelegate {
  /// Creates a delegate that makes grid layouts with a fixed number of tiles in
  /// the cross axis and a fixed main axis extent.
  ///
  /// All of the arguments must not be null. The `mainAxisSpacing` and
  /// `crossAxisSpacing` arguments must not be negative. The `crossAxisCount`
  /// and `mainAxisExtent` arguments must be greater than zero.
  const SliverGridDelegateWithFixedCrossAxisCountAndMainAxisExtent({
    required this.crossAxisCount,
    required this.mainAxisExtent,
    this.mainAxisSpacing = 0.0,
    this.crossAxisSpacing = 0.0,
  }) : assert(crossAxisCount != null && crossAxisCount > 0),
       assert(mainAxisSpacing != null && mainAxisSpacing >= 0),
       assert(crossAxisSpacing != null && crossAxisSpacing >= 0),
       assert(mainAxisExtent != null && mainAxisExtent > 0);

  /// The number of children in the cross axis.
  final int crossAxisCount;

  /// The number of logical pixels between each child along the main axis.
  final double mainAxisSpacing;

  /// The number of logical pixels between each child along the cross axis.
  final double crossAxisSpacing;

  /// The main-axis extent of each child.
  final double mainAxisExtent;

  bool _debugAssertIsValid() {
    assert(crossAxisCount > 0);
    assert(mainAxisSpacing >= 0.0);
    assert(crossAxisSpacing >= 0.0);
    assert(mainAxisExtent > 0.0);
    return true;
  }

  @override
  SliverGridLayout getLayout(SliverConstraints constraints) {
    assert(_debugAssertIsValid());
    final double usableCrossAxisExtent = math.max(0.0,
        constraints.crossAxisExtent - crossAxisSpacing * (crossAxisCount - 1));
    final double childCrossAxisExtent = usableCrossAxisExtent / crossAxisCount;
    final double childMainAxisExtent = mainAxisExtent;
    return SliverGridRegularTileLayout(
      crossAxisCount: crossAxisCount,
      mainAxisStride: childMainAxisExtent + mainAxisSpacing,
      crossAxisStride: childCrossAxisExtent + crossAxisSpacing,
      childMainAxisExtent: childMainAxisExtent,
      childCrossAxisExtent: childCrossAxisExtent,
      reverseCrossAxis: axisDirectionIsReversed(constraints.crossAxisDirection),
    );
  }

  @override
  bool shouldRelayout(SliverGridDelegateWithFixedCrossAxisCountAndMainAxisExtent oldDelegate) {
    return oldDelegate.crossAxisCount != crossAxisCount
        || oldDelegate.mainAxisSpacing != mainAxisSpacing
        || oldDelegate.crossAxisSpacing != crossAxisSpacing
        || oldDelegate.mainAxisExtent != mainAxisExtent;
  }
}
```

### PR DESCRIPTION
**Pull request by chinmoy12c**
## Description

Added mainAxisExtent to SliverGridDelegateWithFixedCrossAxisCount and SliverGridDelegateWithMaxCrossAxisExtent.

## Related Issues

Fixes #71282

## Tests

I added the following tests:

- SliverGridDelegateWithFixedCrossAxisCountAndMainAxisExtent works as expected.
- SliverGridDelegateWithMaxCrossAxisExtentAndFixedMainAxisExtent works as expected.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
Hi @chinmoy! Thanks for the contribution!
There is a lot of common code here from `SliverGridDelegateWithFixedCrossAxisCount` and `SliverGridDelegateWithMaxCrossAxisExtent`.

Instead, what do you think about adding `mainAxisExtent` to the already existing delegates, and when set, let it determine the `childMainAxisExtent` instead of using `childAspectRatio`?

I think that would be simpler and reduce code duplication. It would also prevent us from having some very long names here. 😆

**Comment by chinmoy12c** [comment]
> Hi @chinmoy! Thanks for the contribution!
> There is a lot of common code here from `SliverGridDelegateWithFixedCrossAxisCount` and `SliverGridDelegateWithMaxCrossAxisExtent`.
> 
> Instead, what do you think about adding `mainAxisExtent` to the already existing delegates, and when set, let it determine the `childMainAxisExtent` instead of using `childAspectRatio`?
> 
> I think that would be simpler and reduce code duplication. It would also prevent us from having some very long names here.

Hi @Piinks, I myself was concerned over those long names, bless my editor's auto completion for that :satisfied: . I'll make the necessary changes. Also, should I add an assert so that ```mainAxisExtent``` and ```childAspectRatio``` cannot be set together?

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 1]
```diff
@@ -290,14 +290,16 @@ class SliverGridDelegateWithFixedCrossAxisCount extends SliverGridDelegate {
   /// Creates a delegate that makes grid layouts with a fixed number of tiles in
   /// the cross axis.
   ///
-  /// All of the arguments must not be null. The `mainAxisSpacing` and
-  /// `crossAxisSpacing` arguments must not be negative. The `crossAxisCount`
-  /// and `childAspectRatio` arguments must be greater than zero.
+  /// All of the arguments except [mainAxisExtent] must not be null.
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver_grid.dart:293]: This is looking good. Let's add some more documentation here that explains the two different behaviors, using childAspectRatio (the default) and using mainAxisExtent.

Adding a dartpad example of each would be awesome and easy for folks to play with and see the difference in action, There are examples like that in the code that are wrapped in a `{@tool dartpad ...`

File: packages/flutter/lib/src/rendering/sliver_grid.dart [hunk 2]
```diff
@@ -315,6 +317,9 @@ class SliverGridDelegateWithFixedCrossAxisCount extends SliverGridDelegate {
   /// The ratio of the cross-axis to the main-axis extent of each child.
   final double childAspectRatio;
 
+  /// The extent of each tile in the main axis.
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver_grid.dart:320]: Here as well we should explain how this is used and what the default behavior is.

**Comment by Piinks** [comment]
> Also, should I add an assert so that `mainAxisExtent` and `childAspectRatio` cannot be set together?

I don't think we'll need to add an assert since `childAspectRatio` has a default value. Adding documentation will probably be most helpful. :)

File: packages/flutter/lib/src/rendering/sliver_grid.dart (hunk 1 above)
Comment (chinmoy12c) [reply packages/flutter/lib/src/rendering/sliver_grid.dart:293]: On it :+1:

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Thank you for the contribution!

**Comment by chinmoy12c** [comment]
> Flutter_LGTM
> Thank you for the contribution!

Thank you 😊 Loving the flutter community ❤️

**Pull request merged**